WordPress/gutenberg
# Convert core-data selectors from jsDoc annotations into TypeScript type signatures

## What?

This is a subset of https://github.com/WordPress/gutenberg/pull/39025 – a mega branch that proposes TypeScript signatures for all `@wordpress/core-data` selectors.

This PR removes the jsDoc annotations like `@param {Object}` and `@return {string}` in favor of their TypeScript counterparts. This is the first step towards introducing exhaustive type signatures after https://github.com/WordPress/gutenberg/pull/40024 is merged.

There are two shortcomings for now:

1. ~The doc tool that handles `START TOKEN(Autogenerated selectors` fails. Apparently this is the first instance of mixing it with TypeScript files in the repo.~ (edit: should be fixed by #40236)
2. `createRegistrySelector` isn't typed at the moment, which means that calling it like `const mySelector = createRegistrySelector( select => ( state: State ) => null )` produces a `mySelector` variable of type `any`. We could solve it here by adding a `.d.ts` file to provide the missing signature, or we could solve this problem in `@wordpress/data` instead.

## Testing Instructions

Confirm the checks are green and that no entity configuration got changed when I split the large array into atomic declarations. The changes here should only affect the type system so there is nothing to test in the browser.

cc @dmsnell @jsnajdr @youknowriad @sarayourfriend @getdave @draganescu @scruffian
